### PR TITLE
TandemFoil: cosine LR cycle sweep (t_max=10,20,30,50) with physics+AdamW

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

With the default cosine_t_max=150 and only 4-5 epochs in the 30-minute TandemFoil budget, the cosine LR schedule barely anneals (we start near lr_max and end ~lr_max). Shorter cycles (cosine_t_max=10–30) complete full annealing cycles and may significantly accelerate convergence. We test 4 values with the winning physics + AdamW combination from this round.

## Current Baseline

- **val_primary/surface_pressure_mae: 262.82** (PR #2414, tanjiro — physics features + Lion lr=3e-4, slices=96, 2 epochs)
- **test_primary/surface_pressure_mae: 257.51**

## Instructions

Run 4 parallel trials with `CUDA_VISIBLE_DEVICES`. Use `--wandb_group radford-tandem-physics-adamw` for all runs. All use physics features + AdamW lr=5e-4, model_slices=64.

Physics flags for all trials: `--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`

### Trial 0 — GPU 0: cosine_t_max=10
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 10 --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-physics-adamw --wandb_name thorfinn/tandem-phys-adamw-tmx10 --agent thorfinn
```

### Trial 1 — GPU 1: cosine_t_max=20
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 20 --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-physics-adamw --wandb_name thorfinn/tandem-phys-adamw-tmx20 --agent thorfinn
```

### Trial 2 — GPU 2: cosine_t_max=30
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 30 --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-physics-adamw --wandb_name thorfinn/tandem-phys-adamw-tmx30 --agent thorfinn
```

### Trial 3 — GPU 3: cosine_t_max=50
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 50 --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-physics-adamw --wandb_name thorfinn/tandem-phys-adamw-tmx50 --agent thorfinn
```

Run all 4 in parallel (background each with `&`). Wait for all to finish.

**Key measurements:**
- val_primary/surface_pressure_mae (best across all epochs) for each trial
- Epochs completed per trial
- Note: val_re_rand is the OOD metric — track it separately

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Summary Table

| Trial | cosine_t_max | Epochs | val_primary/surface_pressure_mae | test_primary/surface_pressure_mae |
|-------|-------------|--------|--------------------------------|----------------------------------|
| **tmx10** | **10** | **2** | **224.74** | **222.73 ★** |
| tmx20 | 20 | 2 | 247.25 | 243.08 |
| tmx30 | 30 | 2 | 270.33 | 263.39 |
| tmx50 | 50 | 1 | 359.27 | 344.61 |
| **Baseline** | **50** (Lion) | **2** | **262.82** | **257.51** |

### Best result: cosine_t_max=10

- **val_primary/surface_pressure_mae: 224.74** (vs baseline 262.82, **-14.5%**)
- **test_primary/surface_pressure_mae: 222.73** (vs baseline 257.51, **-13.5%**)
- Per-split test MAE: single_in_dist=232.90, geom_camber_rc=221.10, geom_camber_cruise=211.70, re_rand=225.20
- W&B run: `nkkinh5n`

### Per-trial details

**Trial 0 — cosine_t_max=10 (GPU 0)** ★ WINNER
- Epoch 1: val=358.44, lr=0.0, loss=0.81
- Epoch 2: val=224.74, lr=5e-4, loss=0.46
- Test: 222.73
- W&B: `nkkinh5n`

**Trial 1 — cosine_t_max=20 (GPU 1)**
- Epoch 1: val=359.24, lr=2.5e-4, loss=0.86
- Epoch 2: val=247.25, lr=0.0, loss=0.46
- Test: 243.08
- W&B: `lzvukqmq`

**Trial 2 — cosine_t_max=30 (GPU 2)**
- Epoch 1: val=393.66, lr=0.0, loss=0.99
- Epoch 2: val=270.33, lr=5e-4, loss=0.49
- Test: 263.39
- W&B: `e0xxem2a`

**Trial 3 — cosine_t_max=50 (GPU 3)**
- Epoch 1: val=359.27, lr=0.0, loss=0.86
- Test: 344.61 (only 1 epoch — timeout at epoch 2 start)
- W&B: `aigi7svr`

### Commands used

All trials: `CUDA_VISIBLE_DEVICES=<gpu> PYTHONUNBUFFERED=1 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine-t-max <value> --model-slices 64 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-physics-adamw --wandb-name thorfinn/tandem-phys-adamw-tmx<value> --agent thorfinn`

### Peak GPU memory

Observed via nvidia-smi during training (4 concurrent processes):
- GPU 0 (tmx10): ~91 GB peak
- GPU 1 (tmx20): ~80 GB peak
- GPU 2 (tmx30): ~88 GB peak
- GPU 3 (tmx50): ~94 GB peak

### What happened

**The hypothesis is strongly confirmed.** Shorter cosine annealing cycles dramatically improve convergence within the 30-minute training budget.

Key observations:
1. **cosine_t_max=10 is the clear winner** with test MAE 222.73 — a 13.5% improvement over the baseline (257.51).
2. **Clear monotonic trend**: shorter cycles → better results (tmx10 > tmx20 > tmx30 > tmx50).
3. With the default cosine_t_max=150 and only 750 training steps per epoch, the LR barely anneals. cosine_t_max=10 completes ~75 full cycles per epoch, ensuring the model repeatedly explores and settles.
4. The tmx50 trial (matching baseline's cosine setting) only got 1 epoch due to timeout, but the 3 trials that got 2 epochs all show massive improvement from epoch 1 to epoch 2.
5. **Note on comparisons**: These trials use AdamW lr=5e-4 + model_slices=64, while the baseline used Lion lr=3e-4 + model_slices=96. The smaller model trains faster (more epochs in budget), but the cosine cycle sweep is valid within this configuration.
6. Running 4 concurrent processes caused severe I/O contention (~28 min/epoch vs ~15 min single-process), limiting all trials to 1-2 epochs.

### Suggested follow-ups

1. **Run tmx10 as a solo process** to get more epochs and see if further improvement is possible with the winning cosine_t_max=10 setting.
2. **Test cosine_t_max=5** — even shorter cycles might help further.
3. **Combine tmx10 with model_slices=96** for a direct comparison with the Lion baseline at full model capacity.
4. **Test cosine warm restarts** (CosineAnnealingWarmRestarts) which might work even better than vanilla CosineAnnealingLR for short cycles.
